### DISCUSSION TITLE
Create Upper Pane that Shows list (`ls`) of Files and Directories of Lower Pane

### DISCUSSION THREAD

**Discussion by brunoherrick** [zellij-org/zellij#2649] [Q&A] [answered]
Dear compatriots,

My goal is to have an upper pane, that continually runs `ls` to present a list of files and folders from the current working directory of the **lower pane**.

Something similar to the layout below:

```
layout {
    pane split_direction="horizontal" {
        pane command="ls" {
            // cwd 
        }        
        pane
    }
}
``` 

![image](https://github.com/zellij-org/zellij/assets/135072213/695382ae-ce23-4877-8539-eec7ab317bf2)

Do you have any idea if this is currently possible and how to implement it? I would truly appreciate your help, thank you!

**Comment by Lockszmith-GH** [comment]
Have you tried using:
```bash
watch ls
```
instead of just ls?

**Reply by brunoherrick** [reply]
Thank you for the suggestion @Lockszmith-GH, however I think it does not work. I have tried the following (I do not recommend testing since it seems to create multiple overlaid panes), but it keeps looping the `ls` over the same directory. 

```
layout {
    pane split_direction="horizontal" {
    pane {
        command "watch"
        args "ls"
    }
        pane
    }
}

```

Probably we need something "internal" such as rust's getcwd() to obtain the current working directory of the other pane.

**Reply by Lockszmith-GH** [reply]
With some trial and error I got this to work:

```kdl
layout {
    pane
    pane split_direction="vertical" {
        pane {
            command "watch"
            args "--color" "ls" "-xFa" "--color" "~/"
        }
    }
}
```

Hope this helps.

You'll probably want to replace ~/ with a different location.
I use `-x` flag for ls, so it will fill the display in columns.

**Reply by brunoherrick** [reply]
Tried your code and it does not work as I expect. I probably wasn't clear, I apologize, the goal is to have the upper pane run "ls" based on the current directory of the other lower pane (not the pane from which `ls` is spawn). 

For instance if in the lower pane I `cd` to home, the upper pane shall run a "ls" in `~`. If I posteriorly `cd` again in the lower pane e.g. to the "Downloads" directory, the upper pane should run a `ls` in "~/Downloads". 

This way we have a birds' eye view of our files at any moment. 

So indeed the problem is at `~/`, we must find a way that this location is continually updated based on the location where I am currently at, in the lower pane.

**Reply by Lockszmith-GH** [reply, marked as answer]
Ah, Interesting.
This is a bit more involved then. Because you need someway to have the `ls` panel, '__communicate__' with the working dir.

The way I would approach this would be by hooking into the prompt, this is updated after every command.
The hook write the output of `pwd` into a file at a designated spot (let's say `~/.local/share/ls-panel-pwd`).
And have the 'ls' panel, using `watch` as above, do an `ls` which grabs the path from that file.

In bash, it would look something like:

Your layout in `~/.local/share/ls-panel-layout.kdl` will look like:

```kdl
layout {
    pane split_direction="horizontal" {
        pane {
            command "watch"
            args "--color" "ls" "-xFa" "--color" "$( cat ~/.local/share/ls-panel-pwd )"
        }

        pane
    }
}
```

In your `.bashrc` add:
```bash
report_pwd_to_file() { pwd > ~/.local/share/ls-panel-pwd; }

PROMPT_COMMAND="report_pwd_to_file; ${PROMPT_COMMAND/report_pwd_to_file; /}"
```

Then to start using this setup, run:

```bash
zellij --layout ~/.local/share/ls-panel-layout.kdl
```

**Reply by brunoherrick** [reply]
Thank you @Lockszmith-GH for your godsend work. I leave herein the equivalent zsh configuration, it might be useful for people that are using it instead of bash:

```
report_pwd_to_file() {
  pwd > ~/.local/share/ls-panel-pwd
}

precmd() {
  report_pwd_to_file
}
```

- Note: In the future this may be improved by implementing some kind of zellij plugin that uses the CWD of a specific pane. The reason is that using `cd` in another terminal window will fortuitously lead to the modification of our `ls` pane, even though in our bottom pane we did not perform any `cd` at all.

**Reply by Lockszmith-GH** [reply]
You are correct, You are getting into multi-process communications here, and this becomes a complex signaling issue.

I think the simplest way would be to have the prompt-hook tied to a toggle, where you switch it on manually inside a shell that you want to track.

You would write 2 funcitons: `auto-ls-on` and `auto-ls-off`, and modify `report_pwd_to_file` a  bit.

`auto-ls-on` writes the current `$PID` (process id) to a lock file.
`auto-ls-off` clears the file.

`report_pwd_to_file`, will update the tracking file, **only** if the PID is matching the number in the lock file.

This will ensure only one PID is updating the file.

Another approach would be to use a local variable insted of a lock file, but this might create a race condition, since it will be possible to have multiple shells updating simultanously - which is less optimal.
